### PR TITLE
Custom model (and XML resources) support

### PR DESCRIPTION
**Pull request by NEstelami**

### CODE REVIEW

File: src/binarytools/FileHelper.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        int32_t fileSize = (int32_t)file.tellg();
+        file.seekg(0);
+        char* data = nullptr;
+        std::vector<uint8_t> result;
+
+        try {
+            data = new char[fileSize];
+            file.read(data, fileSize);
+            result = std::vector<uint8_t>(data, data + fileSize);
+        } catch (const std::exception& e) { delete[] data; }
```
Comment (Kenix3) [inline src/binarytools/FileHelper.h]:
```suggestion
        } catch (const std::exception& e) { delete[] data; throw e; }
```

File: src/binarytools/FileHelper.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        delete[] data;
+
+        return result;
+    };
+
+    static std::string ReadAllText(const fs::path& filePath) {
+        std::ifstream file(filePath, std::ios::in | std::ios::binary | std::ios::ate);
+        int32_t fileSize = (int32_t)file.tellg();
+        file.seekg(0);
+        char* data = new char[fileSize + 1];
```
Comment (Kenix3) [inline src/binarytools/FileHelper.h:45]: Same as above here.

File: include/libultraship/libultra/gbi.h [hunk 3]
```diff
@@ -164,7 +164,13 @@
 #define G_SETFB 0x21
 #define G_RESETFB 0x22
 #define G_SETTIMG_FB 0x23
-#define G_DL_OTR 0x31
+#define G_VTX_OTR2 0x24
```
Comment (Kenix3) [inline include/libultraship/libultra/gbi.h]: Better name than just 2?

File: include/libultraship/libultra/gbi.h [hunk 4]
```diff
@@ -164,7 +164,13 @@
...
+#define G_SETTIMG_OTR2 0x25
```
Comment (Kenix3) [inline include/libultraship/libultra/gbi.h:168]: Better name than just 2?

File: include/libultraship/libultra/gbi.h [hunk 5]
```diff
@@ -164,7 +164,13 @@
...
+#define G_TRI1_OTR 0x26
+#define G_DL_OTR_FILEPATH 0x27
+#define G_PUSHCD 0x28
+#define G_MTX_OTR2 0x29
```
Comment (Kenix3) [inline include/libultraship/libultra/gbi.h:172]: Better name than just 2?

File: include/libultraship/libultra/gbi.h [hunk 6]
```diff
@@ -1726,6 +1732,13 @@ typedef union {
     })
 #define gsSPVertex(v, n, v0) \
     { (_SHIFTL(G_VTX, 24, 8) | _SHIFTL((n), 12, 8) | _SHIFTL((v0) + (n), 1, 7)), (uintptr_t)(v) }
+
+#define gsSPVertexOTR2_P1(filePathPtr) \
```
Comment (Kenix3) [inline include/libultraship/libultra/gbi.h:1736]: Better name than just 2?

File: include/libultraship/libultra/gbi.h [hunk 7]
```diff
@@ -1726,6 +1732,13 @@ typedef union {
...
+    { (_SHIFTL(G_VTX_OTR2, 24, 8), (uintptr_t)(filePathPtr)) }
+
+#define gsSPVertexOTR2_P2(vtxCnt, vtxBufOffset, vtxDataOffset) \
```
Comment (Kenix3) [inline include/libultraship/libultra/gbi.h:1739]: Better name than just 2?

File: src/binarytools/BinaryReader.cpp [hunk 8]
```diff
@@ -192,6 +192,21 @@ std::string Ship::BinaryReader::ReadString() {
     return res;
 }
 
+std::string Ship::BinaryReader::ReadCString() {
+    std::string res;
+
+    unsigned char c = 0;
+    do {
+        if (mStream->GetBaseAddress() >= mStream->GetLength())
```
Comment (Kenix3) [inline src/binarytools/BinaryReader.cpp]: Why not put this in the while? I'm OK with either, just curious on the preference.

Comment (NEstelami) [reply src/binarytools/BinaryReader.cpp]: I usually don't use do-while loops but in this case I felt it made sense. It needs to read at least one character before it knows whether or not to stop (checking for the terminator), a do-while guarantees that. A pure-while would have to be written a bit differently.

**Pull request closed without merging**